File: src/mtglab/decks/validate.py
"""The gate. Nothing is emitted until the deck passes this.

This module exists because of two specific, real mistakes: proposing
Ajani, Nacatl Pariah for a G/W deck (its back face is R/W, so its identity is
illegal) and describing Arahbo's doubling ability as eminence (it is not --
eminence is only the +3/+3, the doubling needs him on the battlefield).

Both are *checkable facts*, and both were missed by reasoning from memory
instead of from the card. So identity and oracle text are now looked up, and
`mtglab decks validate` fails loudly rather than producing a confident,
wrong document.

The gate has exactly one mode switch, and it is `deck.stage` (ADR 13). In a
draft a missing `why` warns; in a curated deck it blocks. Everything else --
legality, colour identity, singleton, size, companion and partner rules -- is a
fact about cards, and facts do not become negotiable because a deck is new. The
stage is read off the deck rather than passed in per call so that there is one
answer to "what is this deck", not a caller's opinion of it.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from mtglab.decks import companion, partners
from mtglab.decks.model import (
    ARCHETYPES,
    CATEGORIES,
    DECK_STAGES,
    DECK_STATUSES,
    THEMES,
    Deck,
)

if TYPE_CHECKING:
    from mtglab.cards.db import CardRecord

SINGLETON_EXEMPT = {
    "plains", "island", "swamp", "mountain", "forest", "wastes",
    "snow-covered plains", "snow-covered island", "snow-covered swamp",
    "snow-covered mountain", "snow-covered forest",
}


@dataclass
class Issue:
    level: str          # "error" blocks generation; "warn" does not
    code: str
    message: str
    card: str | None = None

    def __str__(self) -> str:
        where = f" [{self.card}]" if self.card else ""
        return f"{self.level.upper():<5} {self.code}{where}: {self.message}"


@dataclass
class ValidationReport:
    issues: list[Issue] = field(default_factory=list)

    @property
    def errors(self) -> list[Issue]:
        return [i for i in self.issues if i.level == "error"]

    @property
    def warnings(self) -> list[Issue]:
        return [i for i in self.issues if i.level == "warn"]

    @property
    def ok(self) -> bool:
        return not self.errors

    def add(self, level: str, code: str, message: str, card: str | None = None) -> None:
        self.issues.append(Issue(level, code, message, card))

    def render(self) -> str:
        if not self.issues:
            return "OK -- no issues."
        return "\n".join(str(i) for i in self.issues)


def validate(deck: Deck, cards: dict[str, CardRecord] | None = None, *,
             expected_size: int = 99) -> ValidationReport:
    """Validate a deck. `cards` maps name -> CardRecord from the local pool.

    Passing `cards=None` runs only the structural checks and warns that the
    card-level checks were skipped -- it never silently claims the deck is fine.
    """
    rep = ValidationReport()

    # ---- structure ------------------------------------------------------
    if not deck.commander:
        rep.add("error", "no-commander", "deck has no commander")

    # Structural, so it belongs above the `cards is None` early return -- a
    # nonsense status must not become invisible whenever the pool is missing.
    if deck.status not in DECK_STATUSES:
        rep.add("error", "deck-status",
                f"status {deck.status!r} is not one of {', '.join(DECK_STATUSES)}")

    # An unrecognised stage is an error rather than a silent fallback to
    # `curated`. Falling back would make a typo -- `stage: drafted` -- quietly
    # turn every missing rationale into a blocking error, which is the opposite
    # of what the author asked for and hard to work out from the output.
    if deck.stage not in DECK_STAGES:
        rep.add("error", "deck-stage",
                f"stage {deck.stage!r} is not one of {', '.join(DECK_STAGES)}")
    drafting = deck.stage == "draft"

    # The labelling axes warn rather than block, the same bargain as
    # `unknown-category` below: the warning is for a hand-written file, while
    # an edit through the app chooses from the vocabulary and is refused
    # outright by `edit.set_deck_field`. Absent is fine -- an unlabelled deck
    # is undeclared, not wrong -- but a label outside the vocabulary would
    # silently fragment the very grouping it exists for.
    if deck.archetype and deck.archetype not in ARCHETYPES:
        rep.add("warn", "unknown-archetype",
                f"archetype {deck.archetype!r} is not one of "
                f"{', '.join(ARCHETYPES)}")
    for theme in deck.themes:
        if theme not in THEMES:
            rep.add("warn", "unknown-theme",
                    f"theme {theme!r} is not in the vocabulary; the list "
                    f"grows by editing THEMES in decks/model.py")

    # Two commanders share the command zone, so the deck holds 98 rather than
    # 99. `expected_size` is the single-commander default; adjust it by however
    # many commanders there actually are.
    reasons = []
    if len(deck.commander) > 1:
        expected_size -= len(deck.commander) - 1
        reasons.append(f"{len(deck.commander)} commanders")

    # Yorion is the one companion that changes how big the deck must be
    # ("at least twenty cards more than the minimum deck size"), so the size
    # check has to know about it or a legal Yorion deck fails here. Keyed by
    # name because this runs before the pool is consulted, and because deck
    # size is a property of the deck rather than a judgement about a card.
    bonus = companion.DECK_SIZE_BONUS.get((deck.companion or "").lower(), 0)
    if bonus:
        expected_size += bonus
        reasons.append(f"+{bonus} for {deck.companion}")

    if len(deck.commander) > 2:
        rep.add("error", "too-many-commanders",
                f"{len(deck.commander)} commanders listed; Commander allows at "
                "most two, and only with a pairing ability")

    if deck.total_cards != expected_size:
        because = f" ({', '.join(reasons)})" if reasons else ""
        rep.add("error", "deck-size",
                f"deck has {deck.total_cards} cards in the 99, "
                f"expected {expected_size}{because}")

    seen: dict[str, int] = {}
    for card in deck.cards:
        key = card.name.lower()
        seen[key] = seen.get(key, 0) + card.qty
    for name, count in seen.items():
        if count > 1 and name not in SINGLETON_EXEMPT:
            rep.add("error", "singleton", f"appears {count} times", name)

    for cmd in deck.commander:
        if cmd.lower() in seen:
            rep.add("error", "commander-in-99",
                    "commander is also listed in the 99", cmd)

    # Rule 4 is not negotiable, but it is also not a wall an imported deck has
    # to climb before anyone can look at it. In a curated deck a missing `why`
    # is a claim the deck cannot support, and it blocks, one error per card.
    #
    # In a draft it is outstanding work, and it collapses to a single counted
    # warning. That is not cosmetic: a fresh 99-card import produces 99 issues
    # that all say the same thing, which is a wall nobody reads and would bury
    # the one error that matters -- ADR 8 requires warnings to stay rare enough
    # to be read. ADR 13's argument is precisely that a number is a better
    # prompt than a wall, so the count is the report and the deck file is the
    # per-card list.
    pending = deck.unjustified
    if drafting and pending:
        shown = ", ".join(c.name for c in pending[:6])
        more = f", and {len(pending) - 6} more" if len(pending) > 6 else ""
        rep.add("warn", "draft-incomplete",
                f"{len(pending)} of {len(deck.cards)} cards still need a `why` "
                f"({shown}{more}). Write them, then set `stage: curated` -- the "
                "gate refuses the promotion while any card is still blank")

    for card in deck.cards:
        if card.category not in CATEGORIES:
            rep.add("warn", "unknown-category",
                    f"category {card.category!r} is not one of {', '.join(CATEGORIES)}",
                    card.name)
        if not card.why.strip() and not drafting:
            rep.add("error", "missing-rationale",
                    "no `why` -- every inclusion must justify itself", card.name)

    # ---- card-level -----------------------------------------------------
    if cards is None:
        rep.add("warn", "unverified",
                "no card pool supplied; identity, legality and text were NOT checked")
        return rep

    all_names = deck.commander + [c.name for c in deck.cards] + \
        [c.name for c in deck.swap_board]
    if deck.companion:
        all_names.append(deck.companion)

    missing = [n for n in all_names if n not in cards]
    for name in missing:
        rep.add("error", "unknown-card",
                "not found in the local pool -- check spelling, or refresh "
                "the Scryfall data if this is a new card", name)

    # Colour identity is derived from the commander(s), so it must resolve.
    cmd_records = [cards[c] for c in deck.commander if c in cards]
    if cmd_records:
        identity: frozenset[str] = frozenset()
        for rec in cmd_records:
            identity |= rec.color_identity
        # A Background is a legal commander only as one of a pair, so whether
        # a card qualifies depends on how many commanders there are.
        paired = len(cmd_records) > 1
        for rec in cmd_records:
            if not partners.can_be_commander(rec, paired=paired):
                if partners.nonlegendary_partner(rec):
                    # The Battlebond trap: the ability is there, the type line
                    # is not. Say so, because "does not say it can be your
                    # commander" reads like a data problem rather than a rule.
                    extra = (" -- a nonlegendary creature can't be your "
                             "commander even with a 'partner with' ability")
                elif not paired and partners.is_background(rec):
                    extra = (" -- a Background is only legal as a second "
                             "commander, and this deck lists one")
                else:
                    extra = ""
                rep.add("error", "not-a-commander",
                        f"type line is {rec.type_line!r} and it does not say "
                        f"it can be your commander{extra}", rec.name)

        if len(cmd_records) == 2:
            problem = partners.check_pair(*cmd_records)
            if problem:
                rep.add("error", "illegal-pairing", problem)

        for card in deck.cards:
            # A fresh name: `rec` above is the loop variable over commander
            # records and is never None, and reusing it here would widen it.
            card_rec = cards.get(card.name)
            if card_rec is None:
                continue
            illegal = card_rec.color_identity - identity
            if illegal:
                rep.add("error", "color-identity",
                        f"identity {{{''.join(sorted(card_rec.color_identity))}}} includes "
                        f"{{{''.join(sorted(illegal))}}}, outside the commander's "
                        f"{{{''.join(sorted(identity)) or 'C'}}}", card.name)
            if not card_rec.legal_commander:
                rep.add("error", "banned",
                        "not legal in Commander", card.name)
            if card.category == "land" and not card_rec.is_land:
                rep.add("warn", "category-mismatch",
                        f"filed under 'land' but type line is {card_rec.type_line!r}",
                        card.name)
            if card.category != "land" and card_rec.is_land:
                rep.add("warn", "category-mismatch",
                        f"is a land but filed under {card.category!r}", card.name)

    # ---- companion ------------------------------------------------------
    if deck.companion:
        _check_companion(deck, cards, identity if cmd_records else None, rep)

    return rep


def _check_companion(deck: Deck, cards: dict[str, CardRecord],
                     identity: frozenset[str] | None,
                     rep: ValidationReport) -> None:
    """Validate the companion itself and its deckbuilding restriction.

    Previously this confirmed the card had a Companion ability and stopped, so
    the restriction -- the entire reason a companion costs you anything -- went
    unchecked, as did the companion's own legality and colour identity.
    """
    name = deck.companion
    if name is None:
        return                      # no companion declared; the caller checks
    rec = cards.get(name)
    if rec is None:
        return                      # already reported as unknown-card above

    if not companion.is_companion(rec):
        rep.add("error", "not-a-companion",
                "listed as companion but has no Companion ability", name)
        return

    if not rec.legal_commander:
        rep.add("error", "companion-banned",
                "not legal in Commander, so it cannot be your companion", name)

    if identity is not None:
        illegal = rec.color_identity - identity
        if illegal:
            rep.add("error", "companion-color-identity",
                    f"identity {{{''.join(sorted(rec.color_identity))}}} includes "
                    f"{{{''.join(sorted(illegal))}}}, outside the commander's "
                    f"{{{''.join(sorted(identity)) or 'C'}}}", name)

    if name.lower() in (c.name.lower() for c in deck.cards):
        rep.add("error", "companion-in-99",
                "the companion sits outside the 100, not in the deck", name)

    # "Your starting deck" includes the commander, and excludes the companion.
    entries = [(c.name, cards[c.name]) for c in deck.cards if c.name in cards]
    entries += [(n, cards[n]) for n in deck.commander if n in cards]

    result = companion.check(name, entries, cards)
    if result.unsupported:
        # Never report a restriction as satisfied when it was never evaluated.
        rep.add("warn", "companion-unchecked",
                f"deckbuilding restriction was NOT verified -- "
                f"{result.unsupported}. Condition: {result.condition or 'unknown'}",
                name)
        return

    if result.violations:
        shown = ", ".join(sorted(result.violations)[:6])
        more = len(result.violations) - 6
        if more > 0:
            shown += f", and {more} more"
        level = "error" if result.exact else "warn"
        detail = "" if result.exact else " (heuristic check -- verify by hand)"
        rep.add(level, "companion-restriction",
                f"{len(result.violations)} card(s) break the companion "
                f"restriction{detail}: {shown}. Condition: {result.condition}",
                name)


def reserved_list(deck: Deck, cards: dict[str, CardRecord]) -> list[str]:
    """Reserved List cards in the deck -- surfaced because it is a standing
    constraint the user toggles per deck, not a fixed rule."""
    return sorted(c.name for c in deck.cards
                  if (r := cards.get(c.name)) is not None and r.reserved)

File: go/internal/gate/validate.go
package gate

import (
	"fmt"
	"sort"
	"strings"

	"github.com/aasquier/sylvan-library/go/internal/deck"
	"github.com/aasquier/sylvan-library/go/internal/pool"
	"github.com/aasquier/sylvan-library/go/internal/reference"
	"github.com/aasquier/sylvan-library/go/internal/wire"
)

// Issue is one finding: "error" blocks generation, "warn" does not.
type Issue struct {
	Level   string  `json:"-"`
	Code    string  `json:"code"`
	Message string  `json:"message"`
	Card    *string `json:"card"`
}

// Report is `validate.ValidationReport`.
type Report struct {
	Issues []Issue
}

// Errors is every blocking issue.
func (r *Report) Errors() []Issue {
	out := []Issue{}
	for _, i := range r.Issues {
		if i.Level == "error" {
			out = append(out, i)
		}
	}
	return out
}

// Warnings is every non-blocking issue.
func (r *Report) Warnings() []Issue {
	out := []Issue{}
	for _, i := range r.Issues {
		if i.Level == "warn" {
			out = append(out, i)
		}
	}
	return out
}

// OK is `ValidationReport.ok`: no errors.
func (r *Report) OK() bool { return len(r.Errors()) == 0 }

func (r *Report) add(level, code, message string, card string) {
	issue := Issue{Level: level, Code: code, Message: message}
	if card != "" {
		c := card
		issue.Card = &c
	}
	r.Issues = append(r.Issues, issue)
}

// DefaultSize is the single-commander 99.
const DefaultSize = 99

func braces(colors map[string]bool) string {
	keys := make([]string, 0, len(colors))
	for c := range colors {
		keys = append(keys, c)
	}
	sort.Strings(keys)
	return "{" + strings.Join(keys, "") + "}"
}

func bracesOrC(colors map[string]bool) string {
	if len(colors) == 0 {
		return "{C}"
	}
	return braces(colors)
}

func setOf(list []string) map[string]bool {
	out := map[string]bool{}
	for _, s := range list {
		out[s] = true
	}
	return out
}

func minus(a, b map[string]bool) map[string]bool {
	out := map[string]bool{}
	for k := range a {
		if !b[k] {
			out[k] = true
		}
	}
	return out
}

// Validate is `validate.validate`: the gate. `cards` maps name -> record
// from the pool; nil runs only the structural checks and warns that the
// card-level checks were skipped -- it never silently claims the deck is
// fine. `expectedSize` is the single-commander default, adjusted here for a
// second commander and for Yorion.
func Validate(d *deck.Deck, cards map[string]*pool.CardRecord, expectedSize int) *Report {
	rep := &Report{Issues: []Issue{}}
	model := reference.Deck()

	// ---- structure --------------------------------------------------------
	if len(d.Commander) == 0 {
		rep.add("error", "no-commander", "deck has no commander", "")
	}
	if !contains(model.DeckStatuses, d.Status) {
		rep.add("error", "deck-status", fmt.Sprintf("status %s is not one of %s",
			wire.Quote(d.Status), strings.Join(model.DeckStatuses, ", ")), "")
	}
	if !contains(model.DeckStages, d.Stage) {
		rep.add("error", "deck-stage", fmt.Sprintf("stage %s is not one of %s",
			wire.Quote(d.Stage), strings.Join(model.DeckStages, ", ")), "")
	}
	drafting := d.Stage == "draft"

	archetypes := reference.Themes().Archetypes
	if d.LegacyArchetype != "" {
		detail := ""
		if reference.ArchetypeIndex(d.LegacyArchetype) < 0 {
			detail = fmt.Sprintf("archetype %s is not a class the boards know, so it counts for nothing; ",
				wire.Quote(d.LegacyArchetype))
		}
		rep.add("warn", "legacy-archetype", fmt.Sprintf("%s`archetype:` is a legacy key (ADR 37): declare a "+
			"strategy word in `themes` instead -- %s -- and the edit that does will drop this key itself",
			detail, strings.Join(archetypes, ", ")), "")
	}
	for _, theme := range d.Themes {
		if !reference.IsTheme(theme) {
			rep.add("warn", "unknown-theme", fmt.Sprintf("theme %s is not in the vocabulary, so no filter "+
				"will find this deck by it; the deck page's labels editor sets themes from the "+
				"curated list", wire.Quote(theme)), "")
		}
	}

	// The commanders' own records, read here rather than with the rest of the
	// card-level checks because one Rulebreaker clause changes how big the
	// deck is allowed to be, and the size check is above. `cards` may be nil;
	// indexing a nil map is the same "not found" every missing name gets, so
	// no pool means no clauses and the size check below is untouched.
	cmdRecords := []*pool.CardRecord{}
	for _, c := range d.Commander {
		if rec := cards[c]; rec != nil {
			cmdRecords = append(cmdRecords, rec)
		}
	}
	breakers := ReadRulebreakers(cmdRecords)

	reasons := []string{}
	if len(d.Commander) > 1 {
		expectedSize -= len(d.Commander) - 1
		reasons = append(reasons, fmt.Sprintf("%d commanders", len(d.Commander)))
	}
	companion := ""
	if d.Companion != nil {
		companion = *d.Companion
	}
	if bonus := DeckSizeBonus[strings.ToLower(companion)]; bonus != 0 {
		expectedSize += bonus
		reasons = append(reasons, fmt.Sprintf("+%d for %s", bonus, companion))
	}
	if len(d.Commander) > 2 {
		rep.add("error", "too-many-commanders", fmt.Sprintf("%d commanders listed; Commander allows at "+
			"most two, and only with a pairing ability", len(d.Commander)), "")
	}
	// A Rulebreaker clause can lift the ceiling off the deck (Whtz), which
	// turns the number into a floor: too few cards is still an error, too many
	// is the card doing what it says. Nothing here touches the singleton rule,
	// which is a separate rule and still applies to all 200 of them.
	noCeiling := breakers.NoMaximumDeckSize()
	total := d.TotalCards()
	tooFew, tooMany := total < expectedSize, total > expectedSize && !noCeiling
	if tooFew || tooMany {
		because := ""
		if len(reasons) > 0 {
			because = " (" + strings.Join(reasons, ", ") + ")"
		}
		if noCeiling {
			rep.add("error", "deck-size", fmt.Sprintf("deck has %d cards in the 99, expected at least %d%s "+
				"-- the commander's Rulebreaker lifts the maximum, not the minimum",
				total, expectedSize, because), "")
		} else {
			rep.add("error", "deck-size", fmt.Sprintf("deck has %d cards in the 99, expected %d%s",
				total, expectedSize, because), "")
		}
	}

	seen := map[string]int{}
	seenOrder := []string{}
	for _, card := range d.Cards {
		key := strings.ToLower(card.Name)
		if _, there := seen[key]; !there {
			seenOrder = append(seenOrder, key)
		}
		seen[key] += card.Qty
	}
	for _, name := range seenOrder {
		if count := seen[name]; count > 1 && !reference.IsSingletonExempt(name) {
			rep.add("error", "singleton", fmt.Sprintf("appears %d times", count), name)
		}
	}
	for _, cmd := range d.Commander {
		if _, there := seen[strings.ToLower(cmd)]; there {
			rep.add("error", "commander-in-99", "commander is also listed in the 99", cmd)
		}
	}

	pending := d.Unjustified()
	if drafting && len(pending) > 0 {
		shown := []string{}
		for i, c := range pending {
			if i == 6 {
				break
			}
			shown = append(shown, c.Name)
		}
		more := ""
		if len(pending) > 6 {
			more = fmt.Sprintf(", and %d more", len(pending)-6)
		}
		rep.add("warn", "draft-incomplete", fmt.Sprintf("%d of %d cards still need a `why` (%s%s). Write them, "+
			"then set `stage: curated` -- the gate refuses the promotion while any card is still blank",
			len(pending), len(d.Cards), strings.Join(shown, ", "), more), "")
	}

	checkCombos(d, rep)

	for _, card := range d.Cards {
		if !reference.IsCategory(card.Category) {
			rep.add("warn", "unknown-category", fmt.Sprintf("category %s is not one of %s",
				wire.Quote(card.Category), strings.Join(model.Categories, ", ")), card.Name)
		}
		if strings.TrimSpace(card.Why) == "" && !drafting {
			rep.add("error", "missing-rationale", "no `why` -- every inclusion must justify itself", card.Name)
		}
	}

	// ---- card-level -------------------------------------------------------
	if cards == nil {
		rep.add("warn", "unverified", "no card pool supplied; identity, legality and text were NOT checked", "")
		return rep
	}

	allNames := append([]string{}, d.Commander...)
	for _, c := range d.Cards {
		allNames = append(allNames, c.Name)
	}
	for _, c := range d.SwapBoard {
		allNames = append(allNames, c.Name)
	}
	if companion != "" {
		allNames = append(allNames, companion)
	}
	for _, name := range allNames {
		if cards[name] == nil {
			rep.add("error", "unknown-card", "not found in the local pool -- check spelling, or refresh "+
				"the Scryfall data if this is a new card", name)
		}
	}

	// The same question one block over, answered a softer way. A name in the
	// 99 that nobody can look up makes the deck's legality a guess, which is an
	// error; a name in the combos block that nobody can look up costs the entry
	// its picture and nothing else. The deck is still a deck.
	for _, name := range d.ComboNames() {
		if cards[name] == nil {
			rep.add("warn", "combo-unknown-card", "named in the combos but not found in the local "+
				"pool, so it has no card to show -- check the spelling against the card's printed "+
				"name, or refresh the Scryfall data if this is a new card", name)
		}
	}

	var identity map[string]bool
	if len(cmdRecords) > 0 {
		identity = map[string]bool{}
		for _, rec := range cmdRecords {
			for _, c := range rec.ColorIdentity {
				identity[c] = true
			}
		}
		paired := len(cmdRecords) > 1
		for _, rec := range cmdRecords {
			if !CanBeCommander(rec, paired) {
				extra := ""
				if NonlegendaryPartner(rec) {
					extra = " -- a nonlegendary creature can't be your commander even with a 'partner with' ability"
				} else if !paired && IsBackground(rec) {
					extra = " -- a Background is only legal as a second commander, and this deck lists one"
				}
				rep.add("error", "not-a-commander", fmt.Sprintf("type line is %s and it does not say "+
					"it can be your commander%s", wire.Quote(rec.TypeLine), extra), rec.Name)
			}
		}
		if len(cmdRecords) == 2 {
			if problem := CheckPair(cmdRecords[0], cmdRecords[1]); problem != "" {
				rep.add("error", "illegal-pairing", problem, "")
			}
		}
		// A clause nobody could read is said out loud. The identity check
		// below then runs as though the clause were not printed, which will
		// call legal cards illegal -- so the deck is told that is what
		// happened, rather than left to argue with a report that is quietly
		// working from a rule it could not read.
		for _, rb := range breakers.Unread() {
			rep.add("warn", "rulebreaker-unread", fmt.Sprintf("its Rulebreaker was NOT applied -- %s. "+
				"Colour identity below was checked as though the card did not have one, so any card the "+
				"clause makes legal is reported as an error. Clause: %s",
				rb.Unsupported, rb.Text), rb.Commander)
		}
		// Tolabow's clause is a colour the deck picks by playing it; the loop
		// collects what the instants and sorceries actually reach, and the
		// count is judged once, after.
		chosen := []Entry{}
		for _, card := range d.Cards {
			rec := cards[card.Name]
			if rec == nil {
				continue
			}
			if illegal := minus(setOf(rec.ColorIdentity), identity); len(illegal) > 0 {
				switch {
				case breakers.AnyIdentity(rec):
					// The commander says so in as many words: not an error,
					// and not a warning either. A legal card is legal.
				case breakers.OneChosenColor(rec):
					chosen = append(chosen, Entry{Name: card.Name, Rec: rec})
				default:
					rep.add("error", "color-identity", fmt.Sprintf("identity %s includes %s, outside the commander's %s%s",
						braces(setOf(rec.ColorIdentity)), braces(illegal), bracesOrC(identity),
						breakers.Why()), card.Name)
				}
			}
			if !rec.LegalCommander {
				rep.add("error", "banned", "not legal in Commander", card.Name)
			}
			if card.Category == "land" && !rec.IsLand() {
				rep.add("warn", "category-mismatch", fmt.Sprintf("filed under 'land' but type line is %s",
					wire.Quote(rec.TypeLine)), card.Name)
			}
			if card.Category != "land" && rec.IsLand() {
				rep.add("warn", "category-mismatch", fmt.Sprintf("is a land but filed under %s",
					wire.Quote(card.Category)), card.Name)
			}
		}
		if colors, by := breakers.ColorChoice(chosen, identity); len(colors) > 1 {
			reach := []string{}
			for _, c := range colors {
				reach = append(reach, fmt.Sprintf("{%s} (%s)", c, strings.Join(shortList(by[c]), ", ")))
			}
			rep.add("error", "rulebreaker-color-choice", fmt.Sprintf("the commander's Rulebreaker lets these "+
				"cards reach ONE colour past the commander's %s, and this deck reaches %d: %s. Keep the "+
				"colour you want and cut the rest, or move them to the swap board",
				bracesOrC(identity), len(colors), strings.Join(reach, "; ")), "")
		}
	}

	// ---- companion --------------------------------------------------------
	if companion != "" {
		checkCompanion(d, companion, cards, identity, rep)
	}
	return rep
}

// checkCombos reads the combos block against the deck it describes.
//
// **Warnings, every one of them, and that is the decision rather than the
// default.** A combos block is a reading of the deck, not part of it: a deck
// whose catalogue has drifted is a deck with a stale note, and refusing to
// generate its artifacts over one would be the gate holding a primer hostage to
// a paragraph. It is also commandment 2 -- somebody who catalogues their first
// machine and gets four red errors for it has been told they did it wrong, when
// what happened is that they cut a card afterwards.
//
// Nothing here asks whether the Magic is right. Whether two cards actually go
// infinite together is not a question a deterministic check can answer (ADR
// 14), and pretending otherwise would be the gate having an opinion. What it
// checks is whether the entry still matches the deck sitting next to it.
func checkCombos(d *deck.Deck, rep *Report) {
	in99 := map[string]bool{}
	for _, card := range d.Cards {
		in99[strings.ToLower(strings.TrimSpace(card.Name))] = true
	}
	for _, name := range d.Commander {
		in99[strings.ToLower(strings.TrimSpace(name))] = true
	}
	has := func(name string) bool {
		return in99[strings.ToLower(strings.TrimSpace(name))]
	}

	for _, combo := range d.Combos {
		where := combo.Heading()
		for _, piece := range combo.Cards {
			if has(piece) {
				continue
			}
			// The commonest way a catalogue goes stale: the machine was real,
			// and then a piece of it was cut. Named as the piece rather than as
			// the entry, so the row points at the card to put back.
			rep.add("warn", "combo-piece-missing", fmt.Sprintf("is listed as a piece of "+
				"%s but is not in the 99 -- either it was cut and the combo no longer "+
				"assembles, or it belongs in the entry's `needs` as the card this deck "+
				"is still looking for", wire.Quote(where)), piece)
		}
		if !combo.NearMiss() {
			continue
		}
		if has(combo.Needs) {
			rep.add("warn", "combo-needs-in-99", fmt.Sprintf("is marked as the card %s is "+
				"waiting for, but it is already in the deck -- this machine is complete, "+
				"so move it up into the entry's `cards` and drop the trade",
				wire.Quote(where)), combo.Needs)
		}
		if strings.TrimSpace(combo.Cut) == "" {
			// Aaron's rule, checked rather than trusted: a suggestion the deck
			// cannot act on is not a suggestion.
			rep.add("warn", "combo-no-cut", fmt.Sprintf("%s is one card short and does not "+
				"say what would come out for it; a card to bring in is only a suggestion "+
				"once there is a slot for it", wire.Quote(where)), combo.Needs)
		} else if !has(combo.Cut) {
			rep.add("warn", "combo-cut-missing", fmt.Sprintf("is offered as the cut that "+
				"makes room for %s, but it is not in the 99 -- the slot it would free is "+
				"already free", wire.Quote(combo.Needs)), combo.Cut)
		}
	}
}

// checkCompanion is `validate._check_companion`: the companion itself and
// its deckbuilding restriction.
func checkCompanion(d *deck.Deck, name string, cards map[string]*pool.CardRecord,
	identity map[string]bool, rep *Report) {
	rec := cards[name]
	if rec == nil {
		return // already reported as unknown-card
	}
	if !IsCompanion(rec) {
		rep.add("error", "not-a-companion", "listed as companion but has no Companion ability", name)
		return
	}
	if !rec.LegalCommander {
		rep.add("error", "companion-banned", "not legal in Commander, so it cannot be your companion", name)
	}
	// **A Rulebreaker clause deliberately does not reach here.** Every printed
	// one says what "a deck with this commander can have", and a companion is
	// not in the deck -- it waits outside the hundred and is bought into the
	// hand. The reading is also moot in fact: no companion is an Angel, an
	// Aura, a Phyrexian, an Equipment, a land or a seven-drop, so none of the
	// clauses names one. It is written down because the next reader will
	// wonder, and a silence is not an answer.
	if identity != nil {
		if illegal := minus(setOf(rec.ColorIdentity), identity); len(illegal) > 0 {
			rep.add("error", "companion-color-identity", fmt.Sprintf("identity %s includes %s, outside the commander's %s",
				braces(setOf(rec.ColorIdentity)), braces(illegal), bracesOrC(identity)), name)
		}
	}
	for _, c := range d.Cards {
		if strings.EqualFold(c.Name, name) {
			rep.add("error", "companion-in-99", "the companion sits outside the 100, not in the deck", name)
			break
		}
	}
	entries := []Entry{}
	for _, c := range d.Cards {
		if r := cards[c.Name]; r != nil {
			entries = append(entries, Entry{Name: c.Name, Rec: r})
		}
	}
	for _, n := range d.Commander {
		if r := cards[n]; r != nil {
			entries = append(entries, Entry{Name: n, Rec: r})
		}
	}
	result := CheckCompanion(name, entries, cards)
	if result.Unsupported != "" {
		condition := result.Condition
		if condition == "" {
			condition = "unknown"
		}
		rep.add("warn", "companion-unchecked", fmt.Sprintf("deckbuilding restriction was NOT verified -- %s. Condition: %s",
			result.Unsupported, condition), name)
		return
	}
	if len(result.Violations) > 0 {
		sorted := append([]string{}, result.Violations...)
		sort.Strings(sorted)
		shown := sorted
		if len(shown) > 6 {
			shown = shown[:6]
		}
		text := strings.Join(shown, ", ")
		if more := len(sorted) - 6; more > 0 {
			text += fmt.Sprintf(", and %d more", more)
		}
		level, detail := "error", ""
		if !result.Exact {
			level, detail = "warn", " (heuristic check -- verify by hand)"
		}
		rep.add(level, "companion-restriction", fmt.Sprintf("%d card(s) break the companion "+
			"restriction%s: %s. Condition: %s", len(result.Violations), detail, text, result.Condition), name)
	}
}

// shortList is the report's house habit -- name a few, count the rest -- for a
// message that has to fit beside two or three others on one line.
func shortList(names []string) []string {
	if len(names) <= 3 {
		return names
	}
	return append(append([]string{}, names[:3]...), fmt.Sprintf("and %d more", len(names)-3))
}

func contains(list []string, s string) bool {
	for _, item := range list {
		if item == s {
			return true
		}
	}
	return false
}
